### PR TITLE
*: support compiling tidb to wasm

### PR DESCRIPTION
**Pull request by lucklove**
Signed-off-by: lucklove <gnu.crazier@gmail.com>

### What problem does this PR solve? 
This PR make it possible to compile tidb to wasm and use tidb in browser.

### What is changed and how it works?
- add a term to interact with tidb core directly instead of mysql protol
- add wasm platform support for some low-level operation source file
- add a web interface to interact with term
- add polyfill for golang's missing fs operation in wasm

### Check List 

Tests 
 - Manual test

Code changes

Side effects

Related changes

Release note

### CODE REVIEW

**Comment by lucklove** [comment]
This PR works now, but need a little work to do until I remove "WIP".
```
cd wasm
GOOS=js GOARCH=wasm go build -o main.css
python -m SimpleHTTPServer 8000
```

File: util/mathutil/mathutil_js.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
+package mathutil
+
+const MaxUint = ^uint(0) 
+const MinUint = 0 
+const MaxInt = int(MaxUint >> 1) 
+const MinInt = -MaxInt - 1
+
+func MaxUint64(x, y uint64) uint64 {
```
Comment (siddontang) [inline util/mathutil/mathutil_js.go]: please add comments for the export functions and vars

Comment (lucklove) [reply util/mathutil/mathutil_js.go]: Ok, thanks for reminding

File: go.mod [hunk 2]
```diff
@@ -50,7 +50,7 @@ require (
 	github.com/prometheus/client_model v0.0.0-20180712105110-5c3871d89910
```
Comment (5kbpers) [inline go.mod]: seems to need updating `goleveldb` to `7683883`

**Comment by syrusakbary** [comment]
> GOOS=js GOARCH=wasm go build -o main.css

I just inspected https://play.tidb.io and was a bit weird seeing a 13Mb css file `main.css`.
When inspecting, it was actually a wasm file: https://play.tidb.io/main.css

Any reason on naming it `main.css` rather than `main.wasm` ?

File: Makefile [hunk 3]
```diff
@@ -224,6 +225,14 @@ benchdb:
 importer:
 	$(GOBUILD) -ldflags '$(LDFLAGS)' -o bin/importer ./cmd/importer
 
+wasm:
+	mkdir -p wasm-dist
+	GOOS=js GOARCH=wasm $(GOBUILD) -ldflags '$(LDFLAGS)' -o wasm-dist/main.css wasm/*.go
```
Comment (syrusakbary) [inline Makefile]:
```suggestion
	GOOS=js GOARCH=wasm $(GOBUILD) -ldflags '$(LDFLAGS)' -o wasm-dist/main.wasm wasm/*.go
```

File: Makefile [hunk 4]
```diff
@@ -224,6 +225,14 @@ benchdb:
...
+	cp "$(GOROOT)/misc/wasm/wasm_exec.js" wasm-dist/wasm_exec.js
+	cp wasm/favicon.ico wasm-dist/favicon.ico
+	cp wasm/jquery.console.js wasm-dist/jquery.console.js
+	sed 's/82837504/$(shell stat -f%z wasm-dist/main.css)/g' wasm/index.html.tpl > wasm-dist/index.html
```
Comment (syrusakbary) [inline Makefile]:
```suggestion
	sed 's/82837504/$(shell stat -f%z wasm-dist/main.wasm)/g' wasm/index.html.tpl > wasm-dist/index.html
```

File: .gitignore [hunk 5]
```diff
@@ -1,6 +1,8 @@
 bin
 /tidb-server/tidb-server
 /tidb-server/debug
+/wasm-dist
+/wasm/main.css
```
Comment (syrusakbary) [inline .gitignore]:
```suggestion
/wasm/main.wasm
```

File: wasm/index.html.tpl [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    function unimplemented(callback) {
+        const err = new Error("not implemented");
+        err.code = "ENOSYS";
+        callback(err);
+    }
+    function unimplemented1(_1, callback) { unimplemented(callback); }
+    function unimplemented2(_1, _2, callback) { unimplemented(callback); }
+    function bootstrapGo() {
+        const go = new Go();
+        fetch("main.css")
```
Comment (syrusakbary) [inline wasm/index.html.tpl]:
```suggestion
        fetch("main.wasm")
```

File: wasm/index.html.tpl [hunk 7]
```diff
@@ -0,0 +1,261 @@
+<html>
+
+<head>
+    <meta charset="utf-8" />
+    <title>TiDB Playground</title>
+    <link rel="icon" href="/favicon.ico" type="image/x-icon">
+    <script type="text/javascript" src="https://code.jquery.com/jquery-2.1.1.min.js"></script>
```
Comment (Deardrops) [inline wasm/index.html.tpl]: How about use local static file? For running TiDB wasm in offline environment.

File: wasm/index.html.tpl [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    /* .inner {
+        display: table;
+        margin: 0 auto;
+        font-family: monospace;
+        font-size: 20px;
+    }
+
+    .outer {
+        width: 100%
+    } */
```
Comment (Deardrops) [inline wasm/index.html.tpl]: Please remove these commented css

**Comment by lucklove** [comment]
@syrusakbary thanks for comment. 
We use .css instead of .wasm because many CDNs don't compress .wasm file automatically and unfortunately our CDN is one of them. In fact, the size of main.css is not 13MB, it's about 80MB, after compression, it's 13MB.

**Comment by syrusakbary** [comment]
> We use .css instead of .wasm because many CDNs don't compress .wasm file automatically and unfortunately our CDN is one of them. In fact, the size of main.css is not 13MB, it's about 80MB, after compression, it's 13MB.

Oh no! Thanks for the info :)

**Comment by lucklove** [comment]
@syrusakbary thanks for suggestion. I'll add a comment to makefile

**Comment by lucklove** [comment]
known issue: golang1.13 can't produce large wasm correctly
https://github.com/golang/go/issues/34395

**Comment by 5kbpers** [comment]
This PR is ready to merge, PTAL, thx! @zz-jason @siddontang @ngaut

File: go.mod [hunk 9]
```diff
@@ -78,3 +80,5 @@ require (
 go 1.13
 
 replace github.com/pingcap/check => github.com/tiancaiamao/check v0.0.0-20191119042138-8e73d07b629d
+
+replace github.com/coreos/go-systemd => github.com/5kbpers/go-systemd v0.0.0-20191209150347-994f05092cc6
```
Comment (5kbpers) [inline go.mod]: Note: here we forked `go-systemd` for building WASM.

**Comment by lucklove** [comment]
We have fixed TiDB's Wasm [compatibility problem](https://github.com/pingcap/tidb/pull/14183) and moved this feature to a standalone [repo](https://github.com/pingcap/tidb-wasm), closing this PR.

**Pull request closed without merging**